packages/typed-state/src/index.ts

```diff
@@ -0,0 +1,131 @@
+import {
+	Plugin,
+	ModelEffects,
+	ModelEffectsCreator,
+	ModelReducers,
+	Models,
+} from '@rematch/core'
+
+import { Reducer } from 'redux'
+
+const cachedTypings: Record<string, any> = {}
+
+export type TypedStateConfiguration = {
+	strict?: boolean
+	logSeverity?: 'trace' | 'debug' | 'info' | 'warn' | 'error' | 'fatal'
+}
+
+const DEFAULT_SETTINGS: TypedStateConfiguration = {
+	logSeverity: 'warn',
+	strict: false,
+}
+const PROP_TYPES_SECRET = 'SECRET_DO_NOT_PASS_THIS_OR_YOU_WILL_BE_FIRED'
+
+export function logger(
+	logSeverity: TypedStateConfiguration['logSeverity'],
+	m: string
+): void {
+	if (!logSeverity) return
+	if (logSeverity === 'fatal') {
+		throw Error(m)
+	}
+
+	// eslint-disable-next-line no-console
+	console[logSeverity](m)
+}
+
+function validate(
+	typeSpecs: Record<string, any>,
+	values: Record<string, any>,
+	modelName: string,
+	config: TypedStateConfiguration
+): void {
+	if (process.env.NODE_ENV !== 'production') {
+		const keys = Object.keys(typeSpecs)
+		keys.forEach((typeSpecName) => {
+			if (typeSpecs[typeSpecName]) {
+				const error = typeSpecs[typeSpecName](
+					values,
+					typeSpecName,
+					modelName,
+					'property',
+					null,
+					PROP_TYPES_SECRET
+				)
+
+				if (error instanceof Error && config.logSeverity) {
+					logger(config.logSeverity, `[rematch] ${error.message}`)
+				}
+			}
+		})
+	}
+}
+
+const typedStatePlugin = <
+	TModels extends Models<TModels>,
+	TExtraModels extends Models<TModels> = Record<string, any>
+>(
+	config = DEFAULT_SETTINGS
+): Plugin<TModels, TExtraModels> => {
+	if (!config.logSeverity && config.strict) config.logSeverity = 'warn'
+	if (!config.strict) config.strict = false
+
+	return {
+		exposed: {
+			typings: {},
+		},
+		onModel(model): void {
+			cachedTypings[model.name] = model.typings
+		},
+		createMiddleware: () => (store) => (next) => (action) => {
+			const called = next(action)
+			const [modelName] = action.type.split('/')
+			const typings = cachedTypings[modelName]
+			if (typings) {
+				validate(typings, store.getState()[modelName], modelName, config)
+			} else if (config.strict && config.logSeverity) {
+				logger(
+					config.logSeverity,
+					`[rematch]: Missing typings definitions for \`${modelName}\` model`
+				)
+			}
+			return called
+		},
+	}
+}
+
+export default typedStatePlugin
+
+declare module '@rematch/core' {
```

Comment:
Do you mean this type declaration won't work?

Reply:
From the [TS docs](https://www.typescriptlang.org/docs/handbook/declaration-merging.html#introduction):
> Any number of declarations can be merged; it’s not limited to just two declarations

Reply:
This type declaration won't work. Loses the selectors type, like typescript isn't deep merging.
<img width="615" alt="Screenshot 2021-03-02 at 09 48 31" src="https://user-images.githubusercontent.com/22656541/109622520-7a987c80-7b3c-11eb-820b-b78311278e1f.png">